### PR TITLE
NO-JIRA: deferred updates: fix in-place update handling on reboot

### PR DESCRIPTION
**Pull request by ffromani**
There are conditions on which we should not set the reload flag. This avoid regression in the test
"Profile deferred when applied should trigger changes when applied fist, then deferred when edited, if tuned restart should be kept deferred"

### CODE REVIEW

File: pkg/tuned/controller.go
Comment (MarSik) [inline pkg/tuned/controller.go:1040]: The log still says tuned should be reloaded. I will confuse people working on this in the future.. I think.

Comment (ffromani) [reply pkg/tuned/controller.go:1040]: Well, the flow will ultimately trigger a tuned reload: see lines around 1065.
But still: how can we improve the logging?

Comment (MarSik) [reply pkg/tuned/controller.go:1040]: Ah, right! So this is a fix that updates the fingerprint in addition to setting the reload flag. The reload part is not changing. Ok then.

**Review by MarSik — APPROVED** [review]

File: pkg/tuned/controller.go
Comment (ffromani) [reply pkg/tuned/controller.go:1040]: we do want to _not_ set the reload flag if very specific conditions are met, like the one captured in the test mentioned in the commit message, which is constantly failing without this fix

**Comment by ffromani** [comment]
/hold cancel

managed to deflake further (on my env) but not completely

File: pkg/tuned/controller.go
Comment (jmencak) [inline pkg/tuned/controller.go:1045]: What does `changeRecommend` mean exactly?  I'm assuming it indicates that the recommended profile changed. Is that the case here too?  If not, I'd say we need a better name or a variable.

File: pkg/tuned/controller.go
Comment (jmencak) [inline pkg/tuned/controller.go:1035]: Is it possible to improve the comments/logging for this section?   I'm also a bit uneasy with the complex condition without a comment explaining in what state exactly we've currently landed in.  Ideally, the log message below should also explain it.

**Review by jmencak — COMMENTED** [review]
Thank you for the changes, Francesco.  They look good to me.  I've added some comments which are not exactly fair, because they should've been added them during previous reviews.  Nevertheless, I believe that adding better logging/comments what happens (especially) in the 
```
change.deferredMode == util.DeferUpdate && !inplaceUpdate && c.daemon.recoveredRecommendedProfile == change.recommendedProfile
```
section would help a lot future maintainers.

File: pkg/tuned/controller.go
Comment (ffromani) [reply pkg/tuned/controller.go:1045]: right, in this case is better to set the master flag `reload` directly, rather than later checking the combination of `changeRecommend` and `changeProfiles`

File: pkg/tuned/controller.go
Comment (ffromani) [reply pkg/tuned/controller.go:1035]: added improvements, thanks for raising the point

**Comment by ffromani** [comment]
/hold

want to rerun the e2e testsuite

**Comment by jmencak** [comment]
Thank you for the changes/clarification, Francesco!  I've found some typos and things that (IMO) need further clarification.  I'll post a suggested diff later on and we can work from there.
/hold

**Comment by jmencak** [comment]
I tried to clarify the recently added comments and remove the typos.  Hopefully I didn't introduce any false information/typos.  Feel free to adjust as you see fit and happy to discuss here of over Slack:
```
diff --git a/pkg/tuned/controller.go b/pkg/tuned/controller.go
index e6095297..627f4271 100644
--- a/pkg/tuned/controller.go
+++ b/pkg/tuned/controller.go
@@ -1032,11 +1032,12 @@ func (c *Controller) changeSyncerTuneD(change Change) (synced bool, err error) {
 			c.daemon.recommendedProfile = change.recommendedProfile
 			klog.V(1).Infof("recommended TuneD profile updated from %q to %q [inplaceUpdate=%v nodeRestart=%v]", prevRecommended, change.recommendedProfile, inplaceUpdate, change.nodeRestart)
 
-			// if we get this far, it's either because we detected a node restart or the profile changed (not in-place, thus not-edit).
-			// So first: if this is a edit (replacing, not in-place), we do further logic. If it's a node restart, we just go ahead and process it.
-			// Handling then edits. Id we need to defer only edits, and we are re-processing a profile which we already applied, then we must
-			// not reload tuned, otherwise we're missing the point entirely about deferred updates.
-			// See the test "Profile deferred when applied should trigger changes when applied fist, then deferred when edited, if tuned restart should be kept deferred"
+			// If we get this far, it's either because we detected a node restart or the recommended profile changed.
+			// If it's a node restart, we just process the change. If it's the latter -- TuneD profile switch, i.e. not an in-place update -- we need further
+			// logic to distinguish between the "update" and "always" DeferMode.
+			// For the "update" DeferMode, we only defer profile edits. If we are re-processing a profile which we already applied, then we must
+			// not reload tuned, otherwise we're missing the point of deferred updates.
+			// See the test "Profile deferred when applied should trigger changes when applied first, then deferred when edited, if tuned restart should be kept deferred"
 			// See the commit 3655f22656d4a3aa9f471099305dcd78a9c80320
 			if !inplaceUpdate && change.deferredMode == util.DeferUpdate && c.daemon.recoveredRecommendedProfile == change.recommendedProfile {
 				klog.V(1).Infof("reprocessing profile already in effect; this seems a daemon reload. Skip TuneD reload [deferred=%v recoveredRecommended=%v]", change.deferredMode, c.daemon.recoveredRecommendedProfile)

```

**Comment by jmencak** [comment]
/lgtm
the e2e-gcp-pao and e2e-hypershift-pao failures are unrelated
/retest

**Comment by jmencak** [comment]
Hmm, I think `e2e-gcp-pao` test needs looking into.  I'm getting the same failures (`irqbalance.go`) in https://github.com/openshift/cluster-node-tuning-operator/pull/1163 and also when running the test locally.

**Comment by jmencak** [comment]
Overriding 2 tests because of a known change in TuneD. The tests already passed with the previous version of TuneD and need to be adjusted.

/override ci/prow/e2e-gcp-pao
/override ci/prow/e2e-hypershift-pao

**Comment by yanirq** [comment]
/retitle NO-JIRA: deferred updates: fix in-place update handling on reboot

to be added manually to https://github.com/openshift/cluster-node-tuning-operator/pull/1149

**Pull request merged**
